acceptance_testing.go

```diff
@@ -0,0 +1,834 @@
+	a.driver.Write(t, &want)
+
+	// read all records, but stop acking them after we read half of them
+	var lastPosition Position
+	var lastIndex int
+	for i := 0; i < len(want); i++ {
+		readCtx, cancel := context.WithTimeout(ctx, time.Second*5)
+		defer cancel()
+
+		got, err := source.Read(readCtx)
+		is.NoErr(err)
+
+		if i < len(want)/2 {
+			err = source.Ack(ctx, got.Position)
+			is.NoErr(err)
+			lastPosition = got.Position
+			lastIndex = i
+		}
+	}
+	sourceCleanup()
+
+	// At this point we have read all records but acked only half and stopped
+	// the connector. We should be able to resume from the last record that was
+	// not acked.
+
+	source, sourceCleanup = a.openSource(ctx, t, lastPosition) // listen from position
+	defer sourceCleanup()
+
+	for i := lastIndex + 1; i < len(want); i++ {
+		readCtx, cancel := context.WithTimeout(ctx, time.Second*5)
+		defer cancel()
+
+		got, err := source.Read(readCtx)
+		is.NoErr(err)
+
+		err = source.Ack(ctx, got.Position)
+		is.NoErr(err)
+
+		want[i].Position = got.Position   // position can't be determined in advance
+		want[i].CreatedAt = got.CreatedAt // created at can't be determined in advance
+
+		// TODO do a smarter comparison that checks fields separately (e.g. metadata, position etc.)
+		// TODO ensure position is unique
+		// TODO ensure we check metadata and mark it as successful or skipped if it's not (don't fail test because of metadata)
+		is.Equal(want[i], got)
+	}
+}
+
+func (a acceptanceTest) TestSource_Read_Success(t *testing.T) {
+	a.skipIfNoSource(t)
+	is := is.New(t)
+	ctx := context.Background()
+	defer goleak.VerifyNone(t, a.driver.GoleakOptions(t)...)
+
+	positions := make(map[string]bool)
+	isUniquePosition := func(t *testing.T, p Position) {
+		is := is.New(t)
+		is.True(!positions[string(p)])
+		positions[string(p)] = true
+	}
+
+	// write expectation before source exists
+	want := a.generateRecords(10)
+	a.driver.Write(t, &want)
+
+	source, sourceCleanup := a.openSource(ctx, t, nil) // listen from beginning
+	defer sourceCleanup()
+
+	t.Run("snapshot", func(t *testing.T) {
+		is := is.New(t)
+		for i := 0; i < len(want); i++ {
+			// now try to read from the source
+			readCtx, cancel := context.WithTimeout(ctx, time.Second*5)
+			defer cancel()
+
+			got, err := source.Read(readCtx)
+			is.NoErr(err)
+			isUniquePosition(t, got.Position)
+
+			want[i].Position = got.Position   // position can't be determined in advance
+			want[i].CreatedAt = got.CreatedAt // created at can't be determined in advance
+
+			// TODO do a smarter comparison that checks fields separately (e.g. metadata, position etc.)
+			// TODO ensure position is unique
+			// TODO ensure we check metadata and mark it as successful or skipped if it's not (don't fail test because of metadata)
+			is.Equal(want[i], got)
+		}
+	})
+
+	// while connector is running write more data and make sure the connector
+	// detects it
+	want = a.generateRecords(20)
+	a.driver.Write(t, &want)
```

Comment:
IIUC, this basically limits the detected changes to inserts only?

Reply:
That's right, right now we test inserts only. We might want to check updates and deletes as well, the problem is we don't have a way to represent updated or deleted records right now, but this should be doable once we switch to the OpenCDC record.

Reply:
Cool, that makes sense to me. :+1: